drivers/hwmon/occ.c

```diff
@@ -0,0 +1,1347 @@
+	uint8_t occ_state;
+	uint8_t reserved0;
+	uint8_t reserved1;
+	uint8_t error_log_id;
+	uint32_t error_log_addr_start;
+	uint16_t error_log_length;
+	uint8_t reserved2;
+	uint8_t reserved3;
+	char occ_version_string[17];
+	char sensor_eye_catcher[7];
+	uint8_t num_of_sensor_blocks;
+	uint8_t sensor_data_version;
+	struct sensor_data_block *blocks;
+};
+
+struct occ_response {
+	uint8_t sequence_num;
+	uint8_t cmd_type;
+	uint8_t rtn_status;
+	uint16_t data_length;
+	struct occ_poll_data data;
+	uint16_t chk_sum;
+	int temp_block_id;
+	int freq_block_id;
+	int power_block_id;
+	int caps_block_id;
+};
+
+/* data private to each client */
+struct occ_drv_data {
+	struct i2c_client	*client;
+	struct device		*hwmon_dev;
+	struct mutex		update_lock;
+	bool			valid;
+	unsigned long		last_updated;
+	/* Minimum timer interval for sampling In jiffies */
+	unsigned long		sample_time;
+	struct occ_response	occ_resp;
+};
+
+/* i2c read and write occ sensors */
+
+#define OCC_DATA_MAX 4096 /* 4KB at most */
+#define I2C_STATUS_REG 0x000d0001
+#define I2C_ERROR_REG  0x000d0002
+#define I2C_READ_ERROR 1
+#define I2C_WRITE_ERROR 2
+#define I2C_DATABUFFER_SIZE_ERROR 3
+
+/* To generate attn to OCC */
+#define ATTN_DATA                0x0006B035
+
+/* For BMC to read/write SRAM */
+#define OCB_ADDRESS              0x0006B070
+#define OCB_DATA                 0x0006B075
+#define OCB_STATUS_CONTROL_AND   0x0006B072
+#define OCB_STATUS_CONTROL_OR    0x0006B073
+
+#define OCC_COMMAND_ADDR 0xFFFF6000
+#define OCC_RESPONSE_ADDR 0xFFFF7000
+
+static void deinit_occ_resp_buf(struct occ_response *p)
+{
+	int b;
+
+	if (!p)
+		return;
+
+	if (!p->data.blocks)
+		return;
+
+	for (b = 0; b < p->data.num_of_sensor_blocks; b++) {
+		if (!p->data.blocks[b].sensor)
+			kfree(p->data.blocks[b].sensor);
+		if (!p->data.blocks[b].powr)
+			kfree(p->data.blocks[b].powr);
+		if (!p->data.blocks[b].caps)
+			kfree(p->data.blocks[b].caps);
+	}
+
+	kfree(p->data.blocks);
+
+	memset(p, 0, sizeof(*p));
+}
+
+static ssize_t occ_i2c_read(struct i2c_client *client, char *buf, size_t count)
+{
+	int ret = 0;
+
+	if (count > OCC_DATA_MAX)
+		count = OCC_DATA_MAX;
+
+	dev_dbg(&client->dev, "i2c_read: reading %zu bytes @0x%x.\n",
+		count, client->addr);
+	return i2c_master_recv(client, buf, count);
+}
+
+static ssize_t occ_i2c_write(struct i2c_client *client, const char *buf,
+				size_t count)
+{
+	int ret = 0;
+
+	if (count > OCC_DATA_MAX)
+		count = OCC_DATA_MAX;
+
+	dev_dbg(&client->dev, "i2c_write: writing %zu bytes @0x%x.\n",
+		count, client->addr);
+	return i2c_master_send(client, buf, count);
+}
+
+/* read two 4-byte value */
+static int occ_getscom(struct i2c_client *client, uint32_t address,
+			uint32_t *value0, uint32_t *value1)
+{
+	uint32_t ret = 0;
+	char buf[8];
+
+	/* P8 i2c slave requires address to be shifted by 1 */
+	address = address << 1;
+
+	ret = occ_i2c_write(client, (const char *)&address,
+		sizeof(address));
+	/* FIXME: ast i2c driver does not read corret value
```

Comment:
'correct'
also, this means that ret will be unused, this will generate gcc warnings, perhaps remove ret untill the commented code can be uncommented.